include/util/cell_storage.hpp

```diff
@@ -0,0 +1,314 @@
+                                   std::back_inserter(source_boundary),
+                                   [](const auto &cell_and_node) { return cell_and_node.second; });
+                });
+
+            util::for_each_range(
+                level_destination_boundary.begin(),
+                level_destination_boundary.end(),
+                [&](auto begin, auto end) {
+                    BOOST_ASSERT(std::distance(begin, end) > 0);
+
+                    const auto cell_id = begin->first;
+                    BOOST_ASSERT(level_offset + cell_id < cells.size());
+                    auto &cell = cells[level_offset + cell_id];
+                    cell.num_destination_nodes = std::distance(begin, end);
+                    cell.destination_boundary_offset = destination_boundary.size();
+
+                    std::transform(begin,
+                                   end,
+                                   std::back_inserter(destination_boundary),
+                                   [](const auto &cell_and_node) { return cell_and_node.second; });
+                });
+        }
+
+        // Set weight offsets and calculate total storage size
+        WeightOffset weight_offset = 0;
+        for (auto &cell : cells)
+        {
+            cell.weight_offset = weight_offset;
+            weight_offset += cell.num_source_nodes * cell.num_destination_nodes;
+        }
+
+        weights.resize(weight_offset + 1, INVALID_EDGE_WEIGHT);
+    }
+
+    CellStorage(std::vector<EdgeWeight> weights_,
+                std::vector<NodeID> source_boundary_,
+                std::vector<NodeID> destination_boundary_,
+                std::vector<CellData> cells_,
+                std::vector<std::size_t> level_to_cell_offset_)
+        : weights(std::move(weights_)), source_boundary(std::move(source_boundary_)),
+          destination_boundary(std::move(destination_boundary_)), cells(std::move(cells_)),
+          level_to_cell_offset(std::move(level_to_cell_offset_))
+    {
+    }
+
+    ConstCell GetCell(LevelID level, CellID id) const
+    {
+        const auto level_index = LevelIDToIndex(level);
+        BOOST_ASSERT(level_index < level_to_cell_offset.size());
+        const auto offset = level_to_cell_offset[level_index];
+        const auto cell_index = offset + id;
+        BOOST_ASSERT(cell_index < cells.size());
+        return ConstCell{
+            cells[cell_index], weights.data(), source_boundary.data(), destination_boundary.data()};
+    }
+
+    Cell GetCell(LevelID level, CellID id)
+    {
+        const auto level_index = LevelIDToIndex(level);
+        BOOST_ASSERT(level_index < level_to_cell_offset.size());
+        const auto offset = level_to_cell_offset[level_index];
+        const auto cell_index = offset + id;
+        BOOST_ASSERT(cell_index < cells.size());
+        return Cell{
+            cells[cell_index], weights.data(), source_boundary.data(), destination_boundary.data()};
+    }
```

Comment:
The problem I tried to solve here is around read-only views and read-write views (we need both). Even if a `Cell` itself can be `const` it might return non-const iterators to the backing store. All of the methods of `Cell` are always `const`, the view cannot be changed. I don't see how we could use `const` overloading there tor return appropriate iterators in this case?